src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -1513,6 +1554,16 @@ namespace winrt::Microsoft::Terminal::Control::implementation
 
     void ControlCore::_renderEngineSwapChainChanged(const HANDLE handle)
     {
+        // `handle` is a weak ref to a HANDLE that's ultimately owned by the
+        // render engine's own unique_handle. We'll add another ref to it here.
+        // This will make sure that we always have a valid HANDLE to give to
+        // callers of our own SwapChainHandle method, even if the renderer is
+        // currently in the process of discarding this value and creating a new
+        // one. Callers should have already set up the SwapChainChanged
+        // callback, so this all works out.
+        _lastSwapChainHandle.attach(handle);

```

Comment:
The handle you receive isn't "yours" (the called function's), which means you need to clone it first. Hence the usage of `DuplicateHandle` in `RenderEngineSwapChainChanged`, which would be required here as well.

When I originally proposed this, I wasn't really keeping in mind that `ControlCore` doesn't know anything about the UI thread. If it did, you'd could basically do this inside `ControlCore`:
* `DuplicateHandle` (to keep it alive for the duration of the `co_await` to the UI thread)
* `co_await Dispatcher();`
* `_lastSwapChainHandle = std::move(duplicatedHandle);`
* Call `_SwapChainChangedHandlers(...)` - since the call now happens on the main thread already, the callback functions don't need to `DuplicateHandle` themselves and can just pass it straight to the XAML swap chain setter.

But since `ControlCore` can't do that, then the only alternative I see is to wrap the `_lastSwapChainHandle` in a `til::shared_mutex` and then write and read from that, and to:
* `DuplicateHandle` in `ControlCore::_renderEngineSwapChainChanged` (to store it in a `winrt::handle`)
* `DuplicateHandle` in `ControlCore::SwapChainHandle` (because the swap chain handle might get destroyed, while you don't have the mutex locked anymore after returning from the getter call)
* `DuplicateHandle` in `TermControl::RenderEngineSwapChainChanged` (to keep it alive while you switch to the main thread there)

The previous approach with `GetSwapChainHandle()` is about as inconvenient unfortunately, since it requires the exact same modifications, with the only difference where the handle needs to be wrapped in a mutex (`ControlCore` vs. each renderer). There might be another, simpler solution here I'm just not seeing...

Reply:
In your workaround for the atlas engine, you just blew away the swap chain completely. Is that a "simpler" solution than trying to ferry it around?

Like - what if instead of fragging the entire renderer we just made it make a new swap chain after we resumed it?

Reply:
Well, it's not _strictly_ true that ControlCore doesn't know about the UI thread - I think we actually do give it the Dispatcher for the UI thread of the control it's attached to. So I could do that first thing. That does seem easier. 

I suppose I didn't _get_ that [this diff here](https://github.com/microsoft/terminal/commit/3f29b29d56a3b44297ea98737496932cc76238fa#diff-02e96db1905fd3cc71c3cf2477f0e49cd3a5330cc374508f9b7353f73e894354) was just forcing the recreation of a new swapchain. That doesn't seem like a bad idea, but just stashing it does seem easier to me, tbh.